### PR TITLE
Update `RecordSet`

### PR DESCRIPTION
**Pull request by panh99**

### CODE REVIEW

File: src/py/flwr/common/record/recordset.py [hunk 1]
```diff
@@ -15,54 +15,65 @@
...
-    configs: Dict[str, ConfigsRecord] = field(default_factory=dict)
-
-    def set_parameters(self, name: str, record: ParametersRecord) -> None:
-        """Add a ParametersRecord."""
-        self.parameters[name] = record
-
-    def get_parameters(self, name: str) -> ParametersRecord:
-        """Get a ParametesRecord."""
-        return self.parameters[name]
+    """Enhanced RecordSet with a unified and Pythonic interface."""
```
Comment (danieljanes) [inline src/py/flwr/common/record/recordset.py:31]:
```suggestion
    """RecordSet stores groups of parameters, metrics and configs."""
```

File: src/py/flwr/common/record/recordset.py [hunk 2]
```diff
@@ -15,54 +15,65 @@
...
 
-    def del_parameters(self, name: str) -> None:
-        """Delete a ParametersRecord."""
-        del self.parameters[name]
+    _params_records: TypedDict[str, ParametersRecord]
```
Comment (danieljanes) [inline src/py/flwr/common/record/recordset.py]: For consistency?
```suggestion
    _parameters_records: TypedDict[str, ParametersRecord]
```

File: src/py/flwr/common/record/recordset.py [hunk 3]
```diff
@@ -15,54 +15,65 @@
...
+    def __init__(
+        self,
+        parameters_records: Optional[Dict[str, ParametersRecord]] = None,
+        metrics_records: Optional[Dict[str, MetricsRecord]] = None,
+        configs_records: Optional[Dict[str, ConfigsRecord]] = None,
+    ) -> None:
+        def _get_check_fn(__t: Type[T]) -> Callable[[T], None]:
+            def _check_fn(__v: T) -> None:
+                if not isinstance(__v, __t):
+                    raise TypeError(f"Key must be of type str. You passed {type(__v)}.")
```
Comment (danieljanes) [inline src/py/flwr/common/record/recordset.py:46]:
```suggestion
                    raise TypeError(f"Key must be of type `str`, but `{type(__v)}` was passed")
```

Comment (danieljanes) [reply src/py/flwr/common/record/recordset.py:46]: Also, is `str` correct here? Wouldn't it need to be `__t`?

File: src/py/flwr/common/record/recordset.py [hunk 4]
```diff
@@ -15,54 +15,65 @@
...
+            self._metrics_records.update(metrics_records)
+        if configs_records is not None:
+            self._configs_records.update(configs_records)
 
-    def set_configs(self, name: str, record: ConfigsRecord) -> None:
-        """Add a ConfigsRecord."""
-        self.configs[name] = record
+    @property
+    def parameters_records(self) -> TypedDict[str, ParametersRecord]:
+        """Dictionary of ParametersRecord."""
```
Comment (danieljanes) [inline src/py/flwr/common/record/recordset.py:68]:
```suggestion
        """Dictionary holding ParametersRecord instances."""
```

File: src/py/flwr/common/record/recordset.py [hunk 5]
```diff
@@ -15,54 +15,65 @@
...
+        return self._params_records
 
-    def get_configs(self, name: str) -> ConfigsRecord:
-        """Get a ConfigsRecord."""
-        return self.configs[name]
+    @property
+    def metrics_records(self) -> TypedDict[str, MetricsRecord]:
+        """Dictionary of MetricsRecord."""
```
Comment (danieljanes) [inline src/py/flwr/common/record/recordset.py:73]:
```suggestion
        """Dictionary holding MetricsRecord instances."""
```

File: src/py/flwr/common/record/recordset.py [hunk 6]
```diff
@@ -15,54 +15,65 @@
...
+        return self._metrics_records
 
-    def del_configs(self, name: str) -> None:
-        """Delete a ConfigsRecord."""
-        del self.configs[name]
+    @property
+    def configs_records(self) -> TypedDict[str, ConfigsRecord]:
+        """Dictionary of ConfigsRecord."""
```
Comment (danieljanes) [inline src/py/flwr/common/record/recordset.py:78]:
```suggestion
        """Dictionary holding ConfigsRecord instances."""
```

File: src/py/flwr/common/recordset_compat.py [hunk 7]
```diff
@@ -149,13 +149,15 @@ def _fit_or_evaluate_ins_to_recordset(
...
     ins_str = "fitins" if isinstance(ins, FitIns) else "evaluateins"
-    recordset.set_parameters(
-        name=f"{ins_str}.parameters",
-        record=parameters_to_parametersrecord(ins.parameters, keep_input=keep_input),
+    recordset.parameters_records[f"{ins_str}.parameters"] = (
+        parameters_to_parametersrecord(
+            ins.parameters,
+            keep_input,
+        )
     )
```
Comment (danieljanes) [inline src/py/flwr/common/recordset_compat.py:157]:
```suggestion
    parametersrecord = parameters_to_parametersrecord(ins.parameters, keep_input)
    recordset.parameters_records[f"{ins_str}.parameters"] = parametersrecord        
```

File: src/py/flwr/common/record/recordset.py (hunk 3 above)
Comment (panh99) [reply src/py/flwr/common/record/recordset.py]: Thanks. I've updated it.

**Review by danieljanes — APPROVED** [review]

**Pull request merged**